### DISCUSSION TITLE
Web flasher for rnodes

### DISCUSSION THREAD

**Discussion by Erethon** [markqvist/Reticulum#380] [Help & Questions]
This is a discussion to track progress/ideas on if flashing rnodes via a browser is possible. It's a continuation of [this issue from RNode Firmware](https://github.com/markqvist/RNode_Firmware/issues/28).

I used [esp-web-tools](https://github.com/esphome/esp-web-tools) to flash Rnode 1.64 on a T-Beam v1.1 and it seems to have worked fine [1]. However, in order for a Rnode to work we also need the EEPROM to be properly programmed and this is something that `rnodeconf` does after flashing the board. As such, the board is flashed, but it's not working since the EEPROM is defaulting to all `FF` values.

I'm looking at the source code of rnodeconf/rnode-firmware and esp-web-tools currently to figure out:
a) If we can write to the eeprom using esp-web-tools.
b) What information we're writing to the eeprom and what's the minimum info needed for a rnode to work.

[1]: I first had to merge all the files in a single binary using `esptool merge_bin`, create a manifest file, etc.

**Comment by faragher** [comment]
Deeper question: Is there some reason an RNode can't generate missing EEPROM information at boot? I know the signature to verify the source of the firmware would be missing, but I'm otherwise curious as to if this could be sidestepped altogether.

**Reply by markqvist** [reply]
Yes, since some of the EEPROM parameters tell the firmware what transceiver module is installed on the board, frequency ranges, additional optional pin config, etc. So basically, it's used for those things that the firmware just can't "guess" itself.

On top of that, there's a firmware hash for verifying the firmware integrity, a hash verifying the EEPROM configuration, and a a few other signatures.

If the web flasher can send arbitrary serial commands to the RNode after flashing, it should be easy to get it to bootstrap the EEPROM too. The USB protocol has a command for writing arbitrary EEPROM addresses after unlocking it, so it's basically just building the correct EEPROM chunk and sending it off to the device. Once received by the RNode, it will check that the EEPROM checksum is good, and then lock it for further external edits.

**Comment by markqvist** [comment]
> a) If we can write to the eeprom using esp-web-tools.

The `CMD_ROM_WRITE` command (`0x52`) writes individual bytes to EEPROM addresses, so if you can send arbitrary serial data, it should be doable!
